pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_controller.go

```diff
@@ -0,0 +1,235 @@
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	controllerName = "kcp-permissionclaimlabel"
+)
+
+// NewController returns a new controller for handling permission claims for an APIBinding.
+// it will own the AppliedPermissionClaims and will own the accepted permision claim condition.
+func NewController(
+	kcpClusterClient kcpclient.Interface,
+	dynamicClusterClient dynamic.Interface,
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory,
+	apiBindingInformer apisinformers.APIBindingInformer,
+	apiExportInformer apisinformers.APIExportInformer,
+) (*controller, error) {
+	logger := logging.WithReconciler(klog.Background(), controllerName)
+
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+
+	c := &controller{
+		queue:                queue,
+		kcpClusterClient:     kcpClusterClient,
+		dynamicClusterClient: dynamicClusterClient,
+		ddsif:                dynamicDiscoverySharedInformerFactory,
+
+		apiBindingsLister:  apiBindingInformer.Lister(),
+		apiBindingsIndexer: apiBindingInformer.Informer().GetIndexer(),
+
+		getAPIExport: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error) {
+			key := clusters.ToClusterAwareKey(clusterName, name)
+			return apiExportInformer.Lister().Get(key)
+		},
+	}
+
+	apiBindingInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) { c.enqueueAPIBinding(obj, logger, "") },
+		UpdateFunc: func(_, newObj interface{}) {
+			// TODO: In the future, when we should calculate if there is a need for an update.
+			c.enqueueAPIBinding(newObj, logger, "")
+		},
+		DeleteFunc: func(obj interface{}) { c.enqueueAPIBinding(obj, logger, "") },
+	})
+
+	return c, nil
+
+}
+
+// controller reconciles resource labels that make claimed resources visible to an APIExport
+// owner. It labels resources in the intersection of `APIBinding.status.permissionClaims` and
+// `APIBinding.spec.acceptedPermissionClaims`.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	kcpClusterClient     kcpclient.Interface
+	apiBindingsIndexer   cache.Indexer
+	dynamicClusterClient dynamic.Interface
+	ddsif                *informer.DynamicDiscoverySharedInformerFactory
+
+	apiBindingsLister apislisters.APIBindingLister
+	getAPIExport      func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error)
+}
+
+// enqueueAPIBinding enqueues an APIBinding.
+func (c *controller) enqueueAPIBinding(obj interface{}, logger logr.Logger, reason string) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	if reason != "" {
+		logger.V(2).Info("queueing APIBinding", "key", key, "reason", reason)
+	} else {
+		logger.V(2).Info("queueing APIBinding", "key", key)
```

Comment:
@stevekuznetsov can you brain dump a set of a rules you have been following until now and put them into a gdoc? Would be good to have that to know whether this is the right pattern.

Reply:
i have a better pattern for this. need to come back & apply it here

Reply:
We have `logging.WithQueueKey()` for queueing